### PR TITLE
add fields to generator creation

### PR DESCRIPTION
**Pull request by YenguiSeddik**

### CODE REVIEW

File: src/main/java/org/gridsuite/modification/server/dto/ReactiveCapabilityCurveCreationInfos.java [hunk 1]
```diff
@@ -0,0 +1,25 @@
+package org.gridsuite.modification.server.dto;
```
Comment (FranckLecuyer) [inline src/main/java/org/gridsuite/modification/server/dto/ReactiveCapabilityCurveCreationInfos.java:1]: Missing license and author

File: src/test/java/org/gridsuite/modification/server/utils/MatcherGeneratorCreationInfos.java [hunk 2]
```diff
@@ -41,7 +41,9 @@ public boolean matchesSafely(GeneratorCreationInfos m) {
             || (m.getReactivePowerSetpoint() == null && reference.getReactivePowerSetpoint() == null))
             && m.isVoltageRegulationOn() == reference.isVoltageRegulationOn()
             && ((m.getVoltageSetpoint() != null && m.getVoltageSetpoint().equals(reference.getVoltageSetpoint()))
-            || (m.getVoltageSetpoint() == null && reference.getVoltageSetpoint() == null));
+            || (m.getVoltageSetpoint() == null && reference.getVoltageSetpoint() == null))
+            && ((m.getDroop() != null && m.getDroop().equals(reference.getDroop())))
+            || (m.getDroop() == null && reference.getDroop() == null);
```
Comment (FranckLecuyer) [inline src/test/java/org/gridsuite/modification/server/utils/MatcherGeneratorCreationInfos.java:46]: Missing other new added fields

File: src/main/java/org/gridsuite/modification/server/entities/equipment/creation/ReactiveCapabilityCurveCreationEmbeddable.java [hunk 3]
```diff
@@ -0,0 +1,23 @@
+package org.gridsuite.modification.server.entities.equipment.creation;
```
Comment (FranckLecuyer) [inline src/main/java/org/gridsuite/modification/server/entities/equipment/creation/ReactiveCapabilityCurveCreationEmbeddable.java:1]: Missing license and author

File: src/main/java/org/gridsuite/modification/server/service/NetworkModificationService.java [hunk 4]
```diff
@@ -2242,4 +2292,71 @@ public Mono<Void> updateLineAttachToVoltageLevel(UUID modificationUuid, LineAtta
...
+
+            if (identifiable == null) {
+                throw new NetworkModificationException(REGULATING_TERMINAL_NOT_FOUND);
+            }
+
+            if (identifiable instanceof Injection<?>) {
+                return ((Injection) identifiable).getTerminal();
+            } else if (identifiable instanceof Branch) {
+                return ((Branch) identifiable).getTerminal(generatorCreationInfos.getRegulatingTerminalVlId());
+            } else if (identifiable instanceof BusbarSection) {
```
Comment (FranckLecuyer) [inline src/main/java/org/gridsuite/modification/server/service/NetworkModificationService.java:2353]: BusbarSection and HvdcConverterStation are instances of Injection also

File: src/main/java/org/gridsuite/modification/server/service/NetworkModificationService.java [hunk 5]
```diff
@@ -2242,4 +2292,71 @@ public Mono<Void> updateLineAttachToVoltageLevel(UUID modificationUuid, LineAtta
...
+            case VOLTAGE_LEVEL:
+                return network.getVoltageLevel(equipmentId);
+            case BUSBAR_SECTION:
+                return network.getBusbarSection(equipmentId);
+            default:
+                return null;
+        }
+    }
+
+    private Terminal getTerminalFromIdentifiable(GeneratorCreationInfos generatorCreationInfos,
```
Comment (FranckLecuyer) [inline src/main/java/org/gridsuite/modification/server/service/NetworkModificationService.java:2333]: make this method more generic, to be used further with other dtos than GeneratorCreationInfos ...

File: src/main/java/org/gridsuite/modification/server/service/NetworkModificationService.java [hunk 6]
```diff
@@ -976,13 +953,52 @@ private Generator createGeneratorInNodeBreaker(VoltageLevel voltageLevel, Genera
...
+                    .withStepUpTransformerX(generatorCreationInfos.getStepUpTransformerReactance())
+                    .add();
+        }
+
+        if (generatorCreationInfos.getPoints() != null) {
+            generatorCreationInfos.getPoints()
+                    .forEach(point -> generator.newReactiveCapabilityCurve().beginPoint()
+                            .setMaxQ(point.getQmaxP())
+                            .setMinQ(point.getQminP())
+                            .setP(point.getP()));
```
Comment (FranckLecuyer) [inline src/main/java/org/gridsuite/modification/server/service/NetworkModificationService.java:984]: Missing .endPoint() and .add()

File: src/main/java/org/gridsuite/modification/server/service/NetworkModificationService.java [hunk 7]
```diff
@@ -976,13 +953,52 @@ private Generator createGeneratorInNodeBreaker(VoltageLevel voltageLevel, Genera
...
+        }
+
+        if (generatorCreationInfos.getMinimumReactivePower() != null && generatorCreationInfos.getMaximumReactivePower() != null) {
+            generator.newMinMaxReactiveLimits().setMinQ(generatorCreationInfos.getMinimumReactivePower())
+                    .setMaxQ(generatorCreationInfos.getMaximumReactivePower());
```
Comment (FranckLecuyer) [inline src/main/java/org/gridsuite/modification/server/service/NetworkModificationService.java:989]: Missing .add()

File: src/main/java/org/gridsuite/modification/server/service/NetworkModificationService.java [hunk 8]
```diff
@@ -996,6 +1012,40 @@ private Generator createGeneratorInBusBreaker(VoltageLevel voltageLevel, Generat
...
+        }
+
+        if (generatorCreationInfos.getParticipate() != null && generatorCreationInfos.getDroop() != null) {
+            generator.newExtension(ActivePowerControlAdder.class).withParticipate(generatorCreationInfos.getParticipate())
+                    .withDroop(generatorCreationInfos.getDroop());
+        }
+
+        if (generatorCreationInfos.getMaximumReactivePower() != null && generatorCreationInfos.getMinimumReactivePower() != null) {
+            generator.newMinMaxReactiveLimits().setMinQ(generatorCreationInfos.getMinimumReactivePower())
+                    .setMaxQ(generatorCreationInfos.getMaximumReactivePower());
```
Comment (FranckLecuyer) [inline src/main/java/org/gridsuite/modification/server/service/NetworkModificationService.java:1037]: Same remark as above

File: src/main/java/org/gridsuite/modification/server/service/NetworkModificationService.java [hunk 9]
```diff
@@ -996,6 +1012,40 @@ private Generator createGeneratorInBusBreaker(VoltageLevel voltageLevel, Generat
...
+        }
+
+        if (generatorCreationInfos.getPoints() != null) {
+            generatorCreationInfos.getPoints()
+                    .forEach(point -> generator.newReactiveCapabilityCurve().beginPoint()
+                            .setMaxQ(point.getQmaxP())
+                            .setMinQ(point.getQminP())
+                            .setP(point.getP()));
```
Comment (FranckLecuyer) [inline src/main/java/org/gridsuite/modification/server/service/NetworkModificationService.java:1045]: Same remark as above

**Review by FranckLecuyer — CHANGES_REQUESTED** [review]

File: src/main/java/org/gridsuite/modification/server/service/NetworkModificationService.java [hunk 10]
```diff
@@ -859,50 +877,7 @@ private List<EquipmentDeletionInfos> execDeleteEquipment(NetworkStoreListener li
...
-                    case SUBSTATION:
-                        identifiable = network.getSubstation(equipmentId);
-                        break;
-                    case VOLTAGE_LEVEL:
-                        identifiable = network.getVoltageLevel(equipmentId);
-                        break;
-                    default:
-                        break;
-                }
+                Identifiable<?> identifiable = getEquipmentByIdentifiableType(network, equipmentType, equipmentId);
```
Comment (laurgarn) [inline src/main/java/org/gridsuite/modification/server/service/NetworkModificationService.java:880]: Well done, for the factorisation :-)

File: src/main/java/org/gridsuite/modification/server/service/NetworkModificationService.java [hunk 11]
```diff
@@ -2242,4 +2308,62 @@ public Mono<Void> updateLineAttachToVoltageLevel(UUID modificationUuid, LineAtta
...
+            default:
+                return null;
+        }
+    }
+
+    private Terminal getTerminalFromIdentifiable(Network network,
+                                                 String equipmentId,
+                                                 String type,
+                                                 String voltageLevelId) {
+        if (voltageLevelId != null && equipmentId != null && type != null && voltageLevelId != null) {
```
Comment (laurgarn) [inline src/main/java/org/gridsuite/modification/server/service/NetworkModificationService.java]: IntelliJ tells that `voltageLevelId != null` is written twice in this condition.

File: src/main/java/org/gridsuite/modification/server/service/NetworkModificationService.java [hunk 12]
```diff
@@ -2242,4 +2308,62 @@ public Mono<Void> updateLineAttachToVoltageLevel(UUID modificationUuid, LineAtta
...
+                                                 String voltageLevelId) {
+        if (network != null && equipmentId != null && type != null && voltageLevelId != null) {
+            Identifiable<?> identifiable = getEquipmentByIdentifiableType(network, type, equipmentId);
+
+            if (identifiable == null) {
+                throw new NetworkModificationException(EQUIPMENT_NOT_FOUND);
+            }
+
+            if (identifiable instanceof Injection<?>) {
+                return ((Injection) identifiable).getTerminal();
```
Comment (laurgarn) [inline src/main/java/org/gridsuite/modification/server/service/NetworkModificationService.java:2361]: IntelliJ advises to "specialize" / line-up the cast to Injection<?>. The same for `Branch<?>` below.

File: src/test/java/org/gridsuite/modification/server/BuildTest.java [hunk 13]
```diff
@@ -482,7 +493,7 @@ private void testNetworkModificationsCount(UUID groupUuid, int actualSize) {
     public void tearDown() {
         Message<byte[]> message = null;
         try {
-            message = output.receive(TIMEOUT);
+            message = output.receive(4000);
```
Comment (laurgarn) [inline src/test/java/org/gridsuite/modification/server/BuildTest.java:496]: Why quitting from the names constant and change to *more* milliseconds in the teardown ?

File: src/test/java/org/gridsuite/modification/server/utils/MatcherGeneratorCreationInfos.java [hunk 14]
```diff
@@ -41,7 +41,31 @@ public boolean matchesSafely(GeneratorCreationInfos m) {
             || (m.getReactivePowerSetpoint() == null && reference.getReactivePowerSetpoint() == null))
             && m.isVoltageRegulationOn() == reference.isVoltageRegulationOn()
             && ((m.getVoltageSetpoint() != null && m.getVoltageSetpoint().equals(reference.getVoltageSetpoint()))
-            || (m.getVoltageSetpoint() == null && reference.getVoltageSetpoint() == null));
+            || (m.getVoltageSetpoint() == null && reference.getVoltageSetpoint() == null))
+            && ((m.getDroop() != null && m.getDroop().equals(reference.getDroop())))
```
Comment (laurgarn) [inline src/test/java/org/gridsuite/modification/server/utils/MatcherGeneratorCreationInfos.java:45]: What about using `Objects.equals(a, b)` and leaving the `||` ?

**Review by laurgarn — APPROVED** [review]

File: src/test/java/org/gridsuite/modification/server/utils/MatcherGeneratorCreationInfos.java (hunk 14 above)
Comment (YenguiSeddik) [reply src/test/java/org/gridsuite/modification/server/utils/MatcherGeneratorCreationInfos.java]: Done

File: src/test/java/org/gridsuite/modification/server/BuildTest.java (hunk 13 above)
Comment (YenguiSeddik) [reply src/test/java/org/gridsuite/modification/server/BuildTest.java]: unit test fail with that timeout

File: src/main/java/org/gridsuite/modification/server/service/NetworkModificationService.java (hunk 11 above)
Comment (YenguiSeddik) [reply src/main/java/org/gridsuite/modification/server/service/NetworkModificationService.java]: done

File: src/main/java/org/gridsuite/modification/server/service/NetworkModificationService.java (hunk 10 above)
Comment (YenguiSeddik) [reply src/main/java/org/gridsuite/modification/server/service/NetworkModificationService.java]: Thanks

File: src/main/java/org/gridsuite/modification/server/entities/equipment/creation/ReactiveCapabilityCurveCreationEmbeddable.java (hunk 3 above)
Comment (YenguiSeddik) [reply src/main/java/org/gridsuite/modification/server/entities/equipment/creation/ReactiveCapabilityCurveCreationEmbeddable.java:8]: done

File: src/main/java/org/gridsuite/modification/server/dto/ReactiveCapabilityCurveCreationInfos.java (hunk 1 above)
Comment (YenguiSeddik) [reply src/main/java/org/gridsuite/modification/server/dto/ReactiveCapabilityCurveCreationInfos.java:8]: done

File: src/test/java/org/gridsuite/modification/server/utils/MatcherGeneratorCreationInfos.java (hunk 2 above)
Comment (YenguiSeddik) [reply src/test/java/org/gridsuite/modification/server/utils/MatcherGeneratorCreationInfos.java]: done

File: src/main/java/org/gridsuite/modification/server/service/NetworkModificationService.java (hunk 4 above)
Comment (YenguiSeddik) [reply src/main/java/org/gridsuite/modification/server/service/NetworkModificationService.java]: done

File: src/main/java/org/gridsuite/modification/server/service/NetworkModificationService.java (hunk 6 above)
Comment (YenguiSeddik) [reply src/main/java/org/gridsuite/modification/server/service/NetworkModificationService.java]: done

File: src/main/java/org/gridsuite/modification/server/service/NetworkModificationService.java (hunk 7 above)
Comment (YenguiSeddik) [reply src/main/java/org/gridsuite/modification/server/service/NetworkModificationService.java]: I add it

File: src/main/java/org/gridsuite/modification/server/service/NetworkModificationService.java (hunk 8 above)
Comment (YenguiSeddik) [reply src/main/java/org/gridsuite/modification/server/service/NetworkModificationService.java]: done

File: src/main/java/org/gridsuite/modification/server/service/NetworkModificationService.java (hunk 9 above)
Comment (YenguiSeddik) [reply src/main/java/org/gridsuite/modification/server/service/NetworkModificationService.java]: done

**Pull request merged**